### PR TITLE
Making subject field to search type 

### PR DESCRIPTION
**Pull request by rakaar**
- Added a script to scrape all UG and PG courses in `/scripts/scrape_subjects.py` 
- Added a JSON file containing all courses in `/iqps/static/files/code_subjects.json`
- In UploadForm, which is present in `/iqps/upload/views.py`, add a subject with `TypedChoiceField`

Linked issue - #8

### CODE REVIEW

**Comment by rakaar** [comment]
>Also, please specify the reason why is this a "WIP"?

I wanted to test the search dropdown once. I sent the PR for the suggestions regarding scrapped data and data format of subjects.
Like  `CODE - SUBJECT NAME`, this format is okay or any other format?

**Comment by rakaar** [comment]
>Update readme with the required details about the script.

The readme.md currently contains important links and image of the demo, there is no information regarding any script. It would be odd if information regarding  only this script is based
any other places?

**Comment by thealphadollar** [comment]
> > Update readme with the required details about the script.
> 
> The readme.md currently contains important links and image of the demo, there is no information regarding any script. It would be odd if information regarding only this script is based
> any other places?

Yes. We need to start doing it.

**Comment by thealphadollar** [comment]
> > Also, please specify the reason why is this a "WIP"?
> 
> I wanted to test the search dropdown once. I sent the PR for the suggestions regarding scrapped data and data format of subjects.
> Like `CODE - SUBJECT NAME`, this format is okay or any other format?

Ask for forgiveness than permission.

Yes, it looks good to me.

**Comment by rakaar** [comment]
Had bugs in the prev code, fixed this 
Currently like this 
Screenshot from 2020-08-02 20-49-21

**Comment by rakaar** [comment]
i have another small concern-
What if there is a new course added?
Then someone has to ensure that the script is run after every semester and JSON file is updated.
anyway to ensure this @grapheo12 @thealphadollar

**Comment by thealphadollar** [comment]
> i have another small concern-
> What if there is a new course added?
> Then someone has to ensure that the script is run after every semester and JSON file is updated.
> anyway to ensure this @grapheo12 @thealphadollar 

The concern seems valid. 

I can think of two ways to go around it, don't be limited to these.

1. Allow users to select a option that the course is not present in the list and submit without it. In this manner, course will not be a required field.

2. This method is an extension where we would allow users to specify subject code if it is already not in the list governed by the RegEx of subject code. On the backend we would add it directly to the JSON.

On a personal note, I prefer second method since it makes us less worried about subject codes JSON.

I'm aware of the security concerns; however, with the RegEx filtration it becomes very less of a problem.

Please do point out flaws in this method and, recommended, try to think of better ways 🔥

**Comment by rakaar** [comment]
>Allow users to select a option that the course is not present in the list and submit without it. In this manner, course will not be a required field.

If the course is not there, then how will someone search for the paper? 🤔 . It is compulsory

>This method is an extension where we would allow users to specify subject code if it is already not in the list governed by the RegEx of subject code. On the backend we would add it directly to the JSON

Writting to a file directly is not a good idea. Because of concurrent requests, it will erase data. For example for an existing array `Arr`. One request came to append `a`, another req came to append `b`, both reqs came nearly at the same time. Then we either the JSON will have `Arr, a` or `Arr, b`. This is not an unlikely thing. This happened in kwoc 2019.
One more option is save these subjects in Database. But since we are scraping these subjects directly from  site and updating every semester(4 months), updating DB is heavy lifting.

The best option which seems to me is
 Use some bot or something else to remind. Since we are scraping all UG and PG courses from insti site, we should not be missing any courses. But for Just in Case, we can have an input string field, where user can enter manually. And maybe let that subject be updated next semester on our list.

**Comment by thealphadollar** [comment]
> let that subject be updated next semester on our list.

Yes, this seems good. We can also avoid the issue of concurrent writes here by simply using append to the file :D 

> One request came to append a, another req came to append b, both reqs came nearly at the same time. Then we either the JSON will have Arr, a or Arr, b. This is not an unlikely thing. This happened in kwoc 2019.

This is a bit more complex: we can use a simple lock on the file and wait for the previous lock to free. Since we are 100% that we will not be having huge concurrent traffic, this ever this situation comes in one month or two, the lock would be able to handle it easily.

**Comment by rakaar** [comment]
>This is a bit more complex: we can use a simple lock on the file and wait for the previous lock to free. Since we are 100% that we will not be having huge concurrent traffic, this ever this situation comes in one month or two, the lock would be able to handle it easily.

If we are ready to write code for locks and all that stuff. Then I feel saving the subjects in DB is a much simpler option.

The point is not about huge concurrent traffic. In attempt to append new subjects(which are few in number), we are losing a few of them-whats the point?
thats why we better update at the end of the semester or  use DB for subjects too

**Comment by rakaar** [comment]
update after a live discussion
- using a lock, code for which to be written in a separate file
- A separate input field in uploadform for *new* subject addition which checks if the user has entered a valid subject code or not

**Comment by rakaar** [comment]
I have added lock using [filelock](https://github.com/benediktschmitt/py-filelock) package 
but i am facing these two issues
- I have setup docker container and tried installing through `docker-compose run web pip install filelock`, even after that it says `no module named filelock`
- what do i put in the `code_subject.json.lock`

File: iqps/upload/forms.py [hunk 1]
```diff
@@ -1,17 +1,43 @@
...
 from data.models import Paper, Keyword
 from utils.timeutil import current_year
 
 LOG = logging.getLogger(__name__)
 
 
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+FILE_PATH = os.path.join(BASE_DIR, 'static/files/code_subjects.json')
+LOCK_PATH = os.path.join(BASE_DIR, 'static/files/code_subjects.json.lock')
+lock = FileLock(LOCK_PATH, timeout=3)
```
Comment (thealphadollar) [inline iqps/upload/forms.py:20]: let's keep a little higher timeout, say 5 seconds.

I would rather suggest try out adding a new field in the JSON and see how much time it takes. Accordingly, set the timeout to 10x of the observed time.

File: iqps/upload/forms.py [hunk 2]
```diff
@@ -87,6 +120,25 @@ class Meta:
...
+                    with open(FILE_PATH,"w") as f:
+                        data = json.load(f)
+                        code_subjects = data["code_subject"]
+                        code_subjects.append(self.customer)
+                        data = { "code_subject": code_subjects }
+                        json.dump(data, f)
+            except:
+                lock.release()
+            finally:
+                lock.release()
```
Comment (thealphadollar) [inline iqps/upload/forms.py:139]: Finally is always executed whether try encounters an error or not, so it is not required to be kept in except.

Also, in the exception block, we should log the error encountered with a traceback.

**Comment by thealphadollar** [comment]
> * I have setup using docker container and tried installing through `docker-compose run web pip install filelock`, even after that it says `no module named filelock`

Interesting. If I were you, I would try to find the python version the docker file is using and see if the package is available for that version. If it is available, I would let it be installed through `requirements.txt`. Even then if it fails, I'd have to investigate more with error logs etc.

> * what do i put in the `code_subject.json.lock`

Nothing :D It is just there to signify of a lock. It is not supposed to be committed.

File: iqps/upload/forms.py (hunk 1 above)
Comment (rakaar) [reply iqps/upload/forms.py:20]: i tested the time period between reading and writting to the same JSON file
it is in the order of milli seconds most of the times it is 5,6,7 milli seconds
max it was 10 milli seconds.
i stil feel 3 seconds should be reduced to 1 s or 2 second(considering lock will not take too much time)

File: iqps/upload/forms.py (hunk 2 above)
Comment (rakaar) [reply iqps/upload/forms.py]: also logged to info level on new subject being added

**Comment by rakaar** [comment]
>The code_subjects.json.lock is not required to be committed

Does it create the file on its own if not found ?
```
LOCK_PATH = os.path.join(BASE_DIR, 'static/files/code_subjects.json.lock')
lock = FileLock(LOCK_PATH, timeout=3)
```

In future it will cause error right, if someone doesn't create an empty file by that name at that location.Hence I committed that file

**Comment by thealphadollar** [comment]
I'm sure it creates the lock file on its own.

However, I would recommend trying it once and then seeing the result yourself.

**Comment by grapheo12** [comment]
Tested on local machine. It works fine. So I am merging it.
Great work @rakaar 
Thanks for the reviews and the insights @thealphadollar

**Pull request merged**